src/main/scala/is/hail/distributedmatrix/HailBlockMatrix.scala

```diff
@@ -0,0 +1,587 @@
+    m.blocks.iterator(split, context).map { case ((i, j), m) => ((j, i), m.t) }
+
+  protected def getPartitions: Array[Partition] =
+    m.blocks.partitions
+
+  private val prevPartitioner = m.partitioner
+  @transient override val partitioner: Option[Partitioner] =
+    Some(prevPartitioner.transpose)
+}
+
+private class HailBlockMatrixMultiplyRDD(l: HailBlockMatrix, r: HailBlockMatrix)
+  extends RDD[((Int, Int), BDM[Double])](l.blocks.sparkContext, Nil) {
+  require(l.cols == r.rows,
+    s"inner dimension must match, but given: ${ l.rows }x${ l.cols }, ${ r.rows }x${ r.cols }")
+  require(l.blockSize == r.blockSize,
+    s"blocks must be same size, but actually were ${ l.blockSize }x${ l.blockSize } and ${ r.blockSize }x${ r.blockSize }")
+
+  private val lPartitioner = l.partitioner
+  private val lPartitions = l.blocks.partitions
+  private val rPartitioner = r.partitioner
+  private val rPartitions = r.blocks.partitions
+  private val rows = l.rows
+  private val cols = r.cols
+  private val blockSize = l.blockSize
+  private val rowBlocks = lPartitioner.rowPartitions
+  private val colBlocks = rPartitioner.colPartitions
+  private val nProducts = lPartitioner.colPartitions
+  private val rowsRemainder = (rows % blockSize).toInt
+  private val colsRemainder = (cols % blockSize).toInt
+  private val nParts = rowBlocks * colBlocks
+
+  override def getDependencies: Seq[Dependency[_]] =
+    Array[Dependency[_]](
+      new NarrowDependency(l.blocks) {
+        def getParents(partitionId: Int): Seq[Int] = {
+          val row = _partitioner.blockRowIndex(partitionId)
+          val deps = new Array[Int](nProducts)
+          var i = 0
+          while (i < nProducts) {
+            deps(i) = lPartitioner.partitionIdFromBlockIndices(row, i)
+            i += 1
+          }
+          deps
+        }
+      },
+      new NarrowDependency(r.blocks) {
+        def getParents(partitionId: Int): Seq[Int] = {
+          val col = _partitioner.blockColIndex(partitionId)
+          val deps = new Array[Int](nProducts)
+          var i = 0
+          while (i < nProducts) {
+            deps(i) = rPartitioner.partitionIdFromBlockIndices(i, col)
+            i += 1
+          }
+          deps
+        }
+      })
+
+  private def block(hbm: HailBlockMatrix, bmPartitions: Array[Partition], p: HailGridPartitioner, context: TaskContext, i: Int, j: Int): BDM[Double] =
+    hbm.blocks
+      .iterator(bmPartitions(p.partitionIdFromBlockIndices(i, j)), context)
+      .next()
+      ._2
+
+  private def leftBlock(i: Int, j: Int, context: TaskContext): BDM[Double] =
+    block(l, lPartitions, lPartitioner, context, i, j)
+
+  private def rightBlock(i: Int, j: Int, context: TaskContext): BDM[Double] =
+    block(r, rPartitions, rPartitioner, context, i, j)
+
+  def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), BDM[Double])] = {
+    val row = _partitioner.blockRowIndex(split.index)
```

Comment:
I'd use `i` for `row` and `j` for `col` and `k` for `i`.

`row` and `col` are too overloaded, and this reads well:

```
    var k = 0
    while (k < nProducts) {
      val left = leftBlock(i, k, context)
      val right = rightBlock(k, j, context)
      result :+= (left * right)
      k += 1
    }

    Iterator.single(((i, j), result))
```